src/bucket/Bucket.cpp

```diff
@@ -242,7 +258,12 @@ Bucket::loadKeys(
                             }
 #endif
                         }
-                        result.push_back(entryOp->liveEntry());
+
+                        auto const& le = entryOp->liveEntry();
+                        if (!expirationCutoff || isLive(le, *expirationCutoff))
```

Comment:
Hmm. The thing I'm slightly worried about here is that by not-returning an expired entry at bucket level L, we might _reveal_ / revive an earlier not-expired entry at level M>L. Like this changes the a lookup-hit into a lookup-miss, right?

Reply:
I don't think this is possible for a few reasons. First, when we hit the first `DATA_ENTRY` with the key we're searching for, we always remove the key (on line 257), so we stop searching once we hit an expired key. Also, given level M>L, it's not possible for the entry on level M to have a larger expirationLedger than the entry on L since there is no way to make an expirationLedger smaller. This means that an expired entry cannot "shadow" a live entry. This is an interface detail we might change later though, so I can add a test regardless.

Reply:
Actually, it looks like the BucketListDB test "Do not load expired entries" already covers this case. The test shadows valid, non-expired entries with an expired version of the entry and makes sure that the lookup returns null.

Reply:
Hmm, line 257 removes related `EXPIRATION_EXTENSION`s from the search set, but not `DATA_ENTRY` entries.

I'm having trouble interpreting the "do not load expired entries" test (in BucketIndexTests) -- do you mean that one or the "do not load expired entry" test in LedgerTxnTests?

Reply:
(Discussed in chat, satisfied by discussion!)